cpp/src/arrow/filesystem/hdfs.cc

```diff
@@ -0,0 +1,413 @@
+    if (info.kind == io::ObjectType::DIRECTORY) {
+      out->set_type(FileType::Directory);
+      out->set_size(kNoSize);
+    } else if (info.kind == io::ObjectType::FILE) {
+      out->set_type(FileType::File);
+      out->set_size(info.size);
+    }
+    out->set_mtime(ToTimePoint(info.last_modified_time));
+  }
+
+  Status OpenOutputStreamGeneric(const std::string& path, bool append,
+                                 std::shared_ptr<io::OutputStream>* out) {
+    std::shared_ptr<io::HdfsOutputStream> stream;
+    RETURN_NOT_OK(client_->OpenWritable(path, append, options_.buffer_size,
+                                        options_.replication, options_.default_block_size,
+                                        &stream));
+    *out = stream;
+    return Status::OK();
+  }
+
+  bool IsDirectory(const std::string& path) {
+    io::HdfsPathInfo info;
+    Status status = client_->GetPathInfo(path, &info);
+    if (!status.ok()) {
+      return false;
+    }
+    if (info.kind == io::ObjectType::DIRECTORY) {
+      return true;
+    }
+    return false;
+  }
+
+  TimePoint ToTimePoint(int secs) {
+    std::chrono::nanoseconds ns_count(static_cast<int64_t>(secs) * 1000000000);
+    return TimePoint(std::chrono::duration_cast<TimePoint::duration>(ns_count));
+  }
+};
+
+void HdfsOptions::ConfigureEndPoint(const std::string& host, int port) {
+  connection_config.host = host;
+  connection_config.port = port;
+}
+
+void HdfsOptions::ConfigureHdfsDriver(bool use_hdfs3) {
+  if (use_hdfs3) {
+    connection_config.driver = ::arrow::io::HdfsDriver::LIBHDFS3;
+  } else {
+    connection_config.driver = ::arrow::io::HdfsDriver::LIBHDFS;
+  }
+}
+
+void HdfsOptions::ConfigureHdfsUser(const std::string& user_name) {
+  connection_config.user = user_name;
+}
+
+void HdfsOptions::ConfigureHdfsReplication(int16_t replication) {
+  this->replication = replication;
+}
+
+void HdfsOptions::ConfigureHdfsBufferSize(int32_t buffer_size) {
+  this->buffer_size = buffer_size;
+}
+
+void HdfsOptions::ConfigureHdfsBlockSize(int64_t default_block_size) {
+  this->default_block_size = default_block_size;
+}
+
+Status HdfsOptions::FromUri(const Uri& uri, HdfsOptions* out) {
+  *out = HdfsOptions();
+  std::unordered_map<std::string, std::string> options_map;
+  ARROW_ASSIGN_OR_RAISE(const auto options_items, uri.query_items());
+  for (const auto& kv : options_items) {
+    options_map.emplace(kv.first, kv.second);
+  }
+
+  const auto port = uri.port();
+  if (port == -1) {
+    out->ConfigureEndPoint(uri.host(), kDefaultHdfsPort);
+  } else {
+    out->ConfigureEndPoint(uri.host(), port);
+  }
+
+  auto it = options_map.find("use_hdfs3");
+  if (it != options_map.end()) {
+    const auto& v = it->second;
+    if (v == "1") {
+      out->ConfigureHdfsDriver(true);
+    } else if (v == "0") {
+      out->ConfigureHdfsDriver(false);
+    } else {
+      return Status::Invalid("Invalid value for option 'use_hdfs3': '", v, "'");
```

Comment:
Add extra message saying `either use '0' or '1'`, it'll shave some future frustration.

Reply:
Right :-)